openedx/features/badging/helpers.py

```diff
@@ -194,3 +205,73 @@ def send_user_badge_notification(user, my_badge_url, badge_name):
     )
 
     publish_notification_to_user(user.id, message)
+
+
+def get_discussion_team_ids(course_id, discussion_room_id, badges):
+    """
+    Return generic dictionary that contain another dictionary with course_id as key and contain
+    discussion_room_id and team_room_id.
+    :param course_id: Course Id in String format
+    :param discussion_room_id: Discussion room Id in Int format
+    :param badges: Dictionary contain badges
+    """
+    obj = {
+        course_id: {
+            DISCUSSION_ID_KEY: discussion_room_id
+        }
+    }
+    if TEAM_ROOM_ID_KEY in badges:
+        obj[course_id][TEAM_ID_KEY] = badges[TEAM_ROOM_ID_KEY]
+    return obj
+
+
+def get_badge_progress_request_data(username, courses):
+    """
+    Return dictionary that contain username and course data regarding discussion room Id and team room Id.
+    :param username: Username in String format
+    :param courses: course data regarding discussion room Id and team room Id
+    """
+    return {
+        USERNAME_KEY: username,
+        COURSES_KEY: json.dumps(courses)
+    }
+
+
+def add_posts_data_in_badges_list(course, course_key, badges_list):
+    """
+    Add post count data in conversationalist badges and team badges
+    :param course: Dictionary that contain information related to posts count in Discussion room and Team room.
+    :param course_key: Course Id in String format
+    :param badges_list: Dictionary contain badges both conversationalist badges and team badges
+    """
+    discussion_posts_count = course[course_key][DISCUSSION_COUNT_KEY]
+    conversationalist_badges = add_badge_progress(
+        badges_list[CONVERSATIONALIST[CONVERSATIONALIST_ENTRY_INDEX]], discussion_posts_count)
+    badges_list[CONVERSATIONALIST[CONVERSATIONALIST_ENTRY_INDEX]] = conversationalist_badges
+
+    team_posts_count = course[course_key][TEAM_COUNT_KEY] if TEAM_COUNT_KEY in course[course.keys()[0]] else 0
+    team_badges = add_badge_progress(badges_list[TEAM_PLAYER[TEAM_PLAYER_ENTRY_INDEX]], team_posts_count)
+    badges_list[TEAM_PLAYER[TEAM_PLAYER_ENTRY_INDEX]] = team_badges
+
+
+def add_badge_progress(course_badges, posts_count):
+    """
+    Add post count and badge progress in badges that will than rendered in 'my badges' or 'my trophy case'
+    :param course_badges: Badges data either conversationalist badges or team badges
+    :param posts_count: Count of post in Int format
+    """
+    previous_threshold = None
+    for index, badge in enumerate(course_badges):
+        if not previous_threshold:
+            previous_threshold = badge[THRESHOLD_LABEL_KEY]
+        if index > 0:
+            badge[THRESHOLD_LABEL_KEY] = badge[THRESHOLD_LABEL_KEY] - previous_threshold
+            previous_threshold = badge[THRESHOLD_LABEL_KEY] + previous_threshold
```

Comment:
I think we can simplify this as 0 doesn't effect for index 0.
```
previous_threshold = 0
for index, badge in enumerate(course_badges):
	badge[THRESHOLD_LABEL_KEY] = badge[THRESHOLD_LABEL_KEY] - previous_threshold
	previous_threshold = badge[THRESHOLD_LABEL_KEY] + previous_threshold
```

Reply:
We need to assign previous_threshold as threshold of first badge so this is being doing here. the solution you suggested doesn't cater that requirement.

Reply:
This line will do this for you 
```previous_threshold = badge[THRESHOLD_LABEL_KEY] + previous_threshold```
**For first iteration**
badge[THRESHOLD_LABEL_KEY] has threshold of first badge
and previous_threshold is 0, That's why it doesn't effect value when added

Reply:
Done... Good suggestion